### ISSUE TITLE
Allow passing in db interfaces for insert operations 

### ISSUE THREAD

**Issue by Ericnr** [koskimas/kysely#17]
Currently Kysely relies on the user passing in `db.generated` when inserting not null columns with default values since it only knows these columns are not null. If we could pass in separate interfaces for insert operations this process would be more type-safe.

Here's an idea for the api
```ts
interface Person {
  id: number
  first_name: string
  last_name: string
  gender: 'male' | 'female' | 'other'
}

interface Database {
  person: Person
}

interface InsertablePerson {
  // id is generated as identity
  id?: number
  first_name: string
  last_name: string
  gender: 'male' | 'female' | 'other'
}

interface InsertableDatabase {
  person: InsertablePerson
}

// Insertable interfaces are passed to the second type param.
const db = new Kysely<Database, InsertableDatabase>({
  dialect: new PostgresDialect({
    host: 'localhost',
    database: 'kysely_test',
  })
})
```
Some libraries that generate db table interfaces, for example [kanel](https://github.com/kristiandupont/kanel), also generate interfaces for insert operations, which accounts for generated values. It'd be very much useful if we could pass those interfaces to Kysely.

**Comment by koskimas** [comment]
I've considered that, but this would make the types much more complex and things would have one more generic argument everywhere. The tradeoff is not good.

**Comment by koskimas** [comment]
There's some discussion about this idea [here](https://github.com/koskimas/kysely/issues/5)

**Comment by koskimas** [comment]
Reopened this as I'm working on a POC for this after multiple requests.

**Comment by koskimas** [comment]
Here's what I've currently thinking:

```ts
import {
  Kysely,
  PostgresDialect,
  Generated,
  ColumnType,
  Selectable,
  Insertable,
  Updateable,
} from 'kysely'

interface PersonTable {
  // Columns that are generated by the database should be marked
  // using the `Generated` type. This way they are automatically
  // made optional in inserts.
  id: Generated<number>

  first_name: string
  gender: 'male' | 'female' | 'other'

  // If the column is nullable in the database, make its type nullable.
  // Don't use optional properties. Optionality is always determined
  // automatically by Kysely.
  last_name: string | null

  // You can specity the type for each operation (select, insert and update)
  // using the `ColumnType<SelectType, InsertType, UpdateType>` wrapper.
  // Here we define a column `modified_at` that is selected as a `Date`,
  // can optionally be provided as a `string` in inserts and can never
  // be updated:
  modified_at: ColumnType<Date, string | undefined, never>
}

// You can extract the select, insert and update interfaces like this
// if you want (you don't need to):
type Person = Selectable<PersonTable>
type InsertablePerson = Insertable<PersonTable>
type UpdateablePerson = Updateable<PersonTable>

interface PetTable {
  id: Generated<number>
  name: string
  owner_id: number
  species: 'dog' | 'cat'
}

interface MovieTable {
  id: Generated<string>
  stars: number
}

// Keys of this interface are table names.
interface Database {
  person: PersonTable
  pet: PetTable
  movie: MovieTable
}
```

**Comment by koskimas** [comment]
Generated is just a shortcut for `ColumnType`

```ts
export type Generated<S> = ColumnType<S, S | undefined, S>
```

**Comment by koskimas** [comment]
Here's some more documentation

```ts
/**
 * This type can be used to specify a different type for
 * select, insert and update operations.
 *
 * Also see the {@link Generated} type.
 *
 * ### Examples
 *
 * The next example defines a number column that is optional
 * in inserts and updates. All columns are always optional
 * in updates so therefore we don't need to specify `undefined`
 * for the update type. The type below is useful for all kinds of
 * database generated columns like identifiers. The `Generated`
 * type is actually just a shortcut for the type in this example:
 *
 * ```ts
 * ColumnType<number, number | undefined, number>
 * ```
 *
 * The above example makes the column optional in inserts
 * and updates, but you can still choose to provide the
 * column. If you want to prevent insertion/update you
 * can se the type as `never`:
 *
 * ```ts
 * ColumnType<number, never, never>
 * ```
 *
 * Here's one more example where the type is different
 * for each different operation:
 *
 * ```ts
 * ColumnType<Date, string, never>
 * ```
 */
export type ColumnType<
  SelectType,
  InsertType = SelectType,
  UpdateType = SelectType
> = {
  readonly __select__: SelectType
  readonly __insert__: InsertType
  readonly __update__: UpdateType
}

/**
 * A shortcut for defining database-generated columns. The type
 * is the same for all three but when inserted, the value can
 * be omitted.
 *
 * This is implemented using {@link ColumnType} like this:
 *
 * ```ts
 * type Generated<S> = ColumnType<S, S | undefined, S>
 * ```
 */
export type Generated<S> = ColumnType<S, S | undefined, S>

/**
 * Evaluates to `K` if `T` can be `null` or `undefined`.
 */
type IfNullable<T, K> = undefined extends T
  ? K
  : null extends T
  ? K
  : never

/**
 * Evaluates to `K` if `T` can't be `null` or `undefined`.
 */
type IfNotNullable<T, K> = undefined extends T
  ? never
  : null extends T
  ? never
  : T extends never
  ? never
  : K

/**
 * Evaluates to `K` if `T` isn't `never`.
 */
type IfNotNever<T, K> = T extends never ? never : K

export type SelectType<T> = T extends ColumnType<infer S, any, any> ? S : T
export type InsertType<T> = T extends ColumnType<any, infer I, any> ? I : T
export type UpdateType<T> = T extends ColumnType<any, any, infer U> ? U : T

/**
 * Keys of `R` whose `InsertType` values can be `null` or `undefined`.
 */
export type NullableInsertKeys<R> = {
  [K in keyof R]: IfNullable<InsertType<R[K]>, K>
}[keyof R]

/**
 * Keys of `R` whose `InsertType` values can't be `null` or `undefined`.
 */
export type NonNullableInsertKeys<R> = {
  [K in keyof R]: IfNotNullable<InsertType<R[K]>, K>
}[keyof R]

/**
 * Keys of `R` whose `SelectType` values are not `never`
 */
type NonNeverSelectKeys<R> = {
  [K in keyof R]: IfNotNever<SelectType<R[K]>, K>
}[keyof R]

/**
 * Keys of `R` whose `UpdateType` values are not `never`
 */
export type NonNeverUpdateKeys<R> = {
  [K in keyof R]: IfNotNever<UpdateType<R[K]>, K>
}[keyof R]

/**
 * Given a table interface, extracts the select type from all
 * {@link ColumnType} types.
 *
 * ### Examples
 *
 * ```ts
 * interface PersonTable {
 *   id: Generated<number>
 *   first_name: string
 *   modified_at: ColumnType<Date, string, never>
 * }
 *
 * type Person = Selectable<PersonTable>
 * // {
 * //   id: number,
 * //   first_name: string
 * //   modified_at: Date
 * // }
 * ```
 */
export type Selectable<R> = {
  [K in NonNeverSelectKeys<R>]: SelectType<R[K]>
}

/**
 * Given a table interface, extracts the insert type from all
 * {@link ColumnType} types.
 *
 * ### Examples
 *
 * ```ts
 * interface PersonTable {
 *   id: Generated<number>
 *   first_name: string
 *   modified_at: ColumnType<Date, string, never>
 * }
 *
 * type InsertablePerson = Insertable<PersonTable>
 * // {
 * //   id?: number,
 * //   first_name: string
 * //   modified_at: string
 * // }
 * ```
 */
export type Insertable<R> = {
  [K in NonNullableInsertKeys<R>]: InsertType<R[K]>
} & {
  [K in NullableInsertKeys<R>]?: InsertType<R[K]>
}

/**
 * Given a table interface, extracts the update type from all
 * {@link ColumnType} types.
 *
 * ### Examples
 *
 * ```ts
 * interface PersonTable {
 *   id: Generated<number>
 *   first_name: string
 *   modified_at: ColumnType<Date, string, never>
 * }
 *
 * type UpdateablePerson = Updateable<PersonTable>
 * // {
 * //   id?: number,
 * //   first_name?: string
 * // }
 * ```
 */
export type Updateable<R> = {
  [K in NonNeverUpdateKeys<R>]?: UpdateType<R[K]>
}

```

**Comment by koskimas** [comment]
Now implemented in 0.15.0
